filebeat/input/filestream/internal/input-logfile/store.go

```diff
@@ -222,7 +222,10 @@ func (s *sourceStore) FixUpIdentifiers(getNewID func(v Value) (string, interface
 			continue
 		}
 
-		res.lock.Lock()
+		if !res.lock.TryLock() {
```

Comment:
> Is it that the lock is never released? Or is it that the lock is just held too long?

As far as I know those fall into the same bucket. The tests I did with @AndersonQ weren't detailed (or long) enough to differentiate between those two cases.

> How do we know that moving to TryLock won't just prevent entries from being migrated the majority of the time?

I don't know for sure. I'm being optimistic here and using the same implementation logic used by [`CleanIf`](https://github.com/elastic/beats/blob/main/filebeat/input/filestream/internal/input-logfile/store.go#L192-L212) and [`UpdateIdentifiers`](https://github.com/elastic/beats/blob/b3193582de155f89f7ec7da5dfe18a622ed024f7/filebeat/input/filestream/internal/input-logfile/store.go#L260-L294). You can see them all being called on [`prospector.Init`](https://github.com/elastic/beats/blob/b3193582de155f89f7ec7da5dfe18a622ed024f7/filebeat/input/filestream/prospector.go#L63-L125).

In the end I'm taking a best effort approach, Filebeat does it best to migrate the identifiers without blocking the input initialisation forever/for a long time.

Reply:
Thanks, I thinks that's fair. At worst the code is in the same place it was a few weeks ago (no migration happens).

It would be nice if we could understand the root cause here and make this process deterministic eventually.

Reply:
> It would be nice if we could understand the root cause here and make this process deterministic eventually.

I have hunch about it, given two inputs A and B with the same ID, but different files and assuming the Elastic-Agent is sending the config to Filebeat, what we observed was:
1. Input A starts without any problems (aka is harvesting its files)
2. Input B during its initialisation, tries to clean up the registry form its "old files", but because of the ID duplication it tries to clean the file that is being harvested (and locked) by input A. Hence it gets stuck at startup.

I didn't see this behaviour while testing with configuration files so I believe because of the latency between starting different inputs when receiving the configuration from the Elastic-Agent, the input actually has got time to start harvesting the file, hence locking it on the registry.

In the end it all boils down to the same issue: different filestream inputs cannot have the same ID because the ID is the only thing the registry uses to indicate ownership of the files/entries.